File: ios/RunningKitModule.swift
import ExpoModulesCore
import CoreLocation
import CoreMotion

public class RunningKitModule: Module {
    private let locationManager = CLLocationManager()
    private var locationDelegate: RunningKitLocationDelegate?
    
    private let pedometer = CMPedometer()
    private var sessionState = "idle"
    private var workoutStartTime: Date?
    private var totalDistance: Double = 0
    private var lastLocation: CLLocation?
    private var totalSteps: Int = 0
    private var pendingPermissionPromise: Promise?
    private var cadenceTimeoutTimer: Timer?
    private let cadenceTimeoutSeconds: TimeInterval = 3
    private var lastPedometerStepCount: Int = 0
    public func definition() -> ModuleDefinition {
        Name("RunningKit")
        
        Events("onLocationUpdate", "onStepUpdate", "onSessionStateChange")
        
        OnCreate {
            let delegate = RunningKitLocationDelegate(module: self)
            self.locationDelegate = delegate
            self.locationManager.delegate = delegate
            self.locationManager.desiredAccuracy = kCLLocationAccuracyBest
            self.locationManager.distanceFilter = kCLDistanceFilterNone
        }
        
        AsyncFunction("requestPermissions") { (promise: Promise) in
            let status = self.locationManager.authorizationStatus
            if status == .notDetermined {
                self.pendingPermissionPromise = promise
                self.locationManager.requestWhenInUseAuthorization()
            } else {
                promise.resolve(self.buildPermissionsResult(status: status))
            }
        }
        
        AsyncFunction("startWorkout") { (type: String, promise: Promise) in
            guard self.sessionState == "idle" || self.sessionState == "stopped" else {
                promise.reject("ERR_WORKOUT_ACTIVE", "A workout is already in progress")
                return
            }
            self.totalDistance = 0
            self.lastLocation = nil
            self.totalSteps = 0
            self.lastPedometerStepCount = 0
            self.workoutStartTime = Date()
            self.sessionState = "active"
            self.locationManager.startUpdatingLocation()
            self.startPedometerUpdates()
            self.sendEvent("onSessionStateChange", ["state": "active"])
            promise.resolve(nil)
        }
        
        Function("pauseWorkout") {
            guard self.sessionState == "active" else { return }
            self.sessionState = "paused"
            self.locationManager.stopUpdatingLocation()
            self.pedometer.stopUpdates()
            self.cadenceTimeoutTimer?.invalidate()
            self.cadenceTimeoutTimer = nil
            self.sendEvent("onSessionStateChange", ["state": "paused"])
        }
        
        Function("resumeWorkout") {
            guard self.sessionState == "paused" || self.sessionState == "auto-paused" else { return }
            let wasAutoPaused = self.sessionState == "auto-paused"
            self.sessionState = "active"
            self.locationManager.startUpdatingLocation()
            // Pedometer keeps running during auto-pause — restarting it resets currentCadence
            if !wasAutoPaused {
                self.startPedometerUpdates()
            }
            self.sendEvent("onSessionStateChange", ["state": "active"])
        }
        
        AsyncFunction("stopWorkout") { (promise: Promise) in
            guard self.sessionState == "active" || self.sessionState == "paused" || self.sessionState == "auto-paused" else {
                promise.reject("ERR_NO_WORKOUT", "No active workout to stop")
                return
            }
            self.locationManager.stopUpdatingLocation()
            self.pedometer.stopUpdates()
            self.cadenceTimeoutTimer?.invalidate()
            self.cadenceTimeoutTimer = nil
            self.sessionState = "stopped"
            let duration = self.workoutStartTime.map { Date().timeIntervalSince($0) } ?? 0
            self.sendEvent("onSessionStateChange", ["state": "stopped"])
            promise.resolve([
                "duration": duration,
                "distance": self.totalDistance,
                "steps": self.totalSteps,
                "avgSpeed": duration > 0 ? self.totalDistance / duration : 0,
                "calories": Double(self.totalSteps) * 0.04
            ])
        }
        
        
        Function("autoPauseWorkout") {
            guard self.sessionState == "active" else { return }
            self.sessionState = "auto-paused"
            self.locationManager.stopUpdatingLocation()
            self.sendEvent("onSessionStateChange", ["state": "auto-paused"])
        }
    }
    
    // MARK: - Location Handling
    
    func handleLocationUpdate(_ location: CLLocation) {
        guard sessionState == "active" else { return }
        if let last = lastLocation {
            totalDistance += location.distance(from: last)
        }
        lastLocation = location
        sendEvent("onLocationUpdate", [
            "latitude": location.coordinate.latitude,
            "longitude": location.coordinate.longitude,
            "altitude": location.altitude,
            "speed": max(location.speed, 0),
            "accuracy": location.horizontalAccuracy,
            "timestamp": location.timestamp.timeIntervalSince1970 * 1000
        ])
    }
    
    
    // MARK: - Permission Handling
    
    func handleAuthorizationChange(_ status: CLAuthorizationStatus) {
        guard let promise = pendingPermissionPromise else { return }
        promise.resolve(buildPermissionsResult(status: status))
        pendingPermissionPromise = nil
    }
    
    
    // MARK: - Helpers
    
    private func buildPermissionsResult(status: CLAuthorizationStatus) -> [String: String] {
        let locationStatus: String
        switch status {
        case .authorizedAlways, .authorizedWhenInUse: locationStatus = "granted"
        case .denied, .restricted: locationStatus = "denied"
        default: locationStatus = "undetermined"
        }
        return [
            "location": locationStatus,
            "motion": CMPedometer.isStepCountingAvailable() ? "granted" : "denied"
        ]
    }
    
    
    private func startPedometerUpdates() {
        guard CMPedometer.isStepCountingAvailable(), let start = workoutStartTime else { return }

        // startEventUpdates fires almost immediately when walking starts or stops.
        // This gives fast auto-resume without waiting for startUpdates to deliver cadence.
        if CMPedometer.isPedometerEventTrackingAvailable() {
            pedometer.startEventUpdates { [weak self] event, error in
                guard let self, let event, error == nil else { return }
                // Only handle resume — sends instant cadence signal for fast auto-resume.
                // Pause events are deliberately ignored: they fire during normal stride gaps
                // and cause false auto-pause triggers. The cadence timeout handles genuine stops.
                if event.type == .resume {
                    DispatchQueue.main.async { [weak self] in
                        guard let self else { return }
                        self.sendEvent("onStepUpdate", [
                            "steps": self.totalSteps,
                            "cadence": 60.0
                        ])
                    }
                }
            }
        }

        pedometer.startUpdates(from: start) { [weak self] data, error in
            guard let self, let data, error == nil else { return }
            let currentStepCount = data.numberOfSteps.intValue
            let stepsIncreased = currentStepCount > self.lastPedometerStepCount
            self.lastPedometerStepCount = currentStepCount
            self.totalSteps = currentStepCount

            let cadence: Double
            if let raw = data.currentCadence?.doubleValue, raw > 0 {
                cadence = raw * 60
            } else if stepsIncreased {
                cadence = 60.0
            } else {
                cadence = 0
            }

            DispatchQueue.main.async { [weak self] in
                guard let self else { return }
                self.sendEvent("onStepUpdate", ["steps": self.totalSteps, "cadence": cadence])

                // Reset the timeout only when a real new step arrived.
                // currentCadence is a rolling average that stays non-zero for ~3s after stopping,
                // so using it to gate the timer would delay auto-pause significantly.
                if stepsIncreased {
                    self.cadenceTimeoutTimer?.invalidate()
                    let steps = self.totalSteps
                    self.cadenceTimeoutTimer = Timer.scheduledTimer(
                        withTimeInterval: self.cadenceTimeoutSeconds,
                        repeats: false
                    ) { [weak self] _ in
                        guard let self else { return }
                        self.sendEvent("onStepUpdate", ["steps": steps, "cadence": 0])
                    }
                }
            }
        }
    }
    
}

// A separate NSObject subclass is required because CLLocationManagerDelegate inherits
// from NSObjectProtocol, and ExpoModulesCore's Module class does not extend NSObject.
// MARK: - Location Delegate

    class RunningKitLocationDelegate: NSObject, CLLocationManagerDelegate {
        weak var module: RunningKitModule?
        
        init(module: RunningKitModule) {
            self.module = module
        }
        
        func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
            guard let location = locations.last else { return }
            module?.handleLocationUpdate(location)
        }
        
        func locationManagerDidChangeAuthorization(_ manager: CLLocationManager) {
            module?.handleAuthorizationChange(manager.authorizationStatus)
        }
    }

File: android/src/main/java/expo/modules/runningkit/RunningKitModule.kt
package expo.modules.runningkit

import android.content.Context
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import android.os.Handler
import android.os.Looper
import com.google.android.gms.location.*
import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class RunningKitModule : Module() {

  // MARK: - Properties
  private lateinit var fusedLocationClient: FusedLocationProviderClient
  private lateinit var sensorManager: SensorManager
  private var stepSensor: Sensor? = null
  private var sessionState = "idle"
  private var workoutStartTime: Long = 0
  private var totalDistance: Double = 0.0
  private var lastLocation: android.location.Location? = null
  private var totalSteps: Int = 0
  private var initialStepCount: Int = -1

  // Cadence calculation
  private var lastStepCount: Int = 0
  private var lastStepTime: Long = 0
  private var cadenceWindowReset = true

  // Cadence timeout — fires cadence=0 if no steps arrive within threshold
  private val cadenceHandler = Handler(Looper.getMainLooper())
  private val cadenceTimeoutMs = 4000L
  private val cadenceTimeoutRunnable = Runnable {
    if (sessionState == "active" || sessionState == "auto-paused") {
      cadenceWindowReset = true
      sendEvent("onStepUpdate", mapOf(
        "steps" to totalSteps,
        "cadence" to 0
      ))
    }
  }

  // MARK: - Location Callback
  private val locationCallback = object : LocationCallback() {
    override fun onLocationResult(result: LocationResult) {
      val location = result.lastLocation ?: return
      if (sessionState != "active") return

      lastLocation?.let { totalDistance += it.distanceTo(location) }
      lastLocation = location

      sendEvent("onLocationUpdate", mapOf(
        "latitude" to location.latitude,
        "longitude" to location.longitude,
        "altitude" to location.altitude,
        "speed" to maxOf(location.speed.toDouble(), 0.0),
        "accuracy" to location.accuracy.toDouble(),
        "timestamp" to location.time.toDouble()
      ))
    }
  }

  // MARK: - Step Listener
  private val stepListener = object : SensorEventListener {
    override fun onSensorChanged(event: SensorEvent) {
      if (sessionState != "active" && sessionState != "auto-paused") return

      val stepCount = event.values[0].toInt()
      if (initialStepCount == -1) {
        initialStepCount = stepCount
        lastStepCount = stepCount
        lastStepTime = System.currentTimeMillis()
      }

      totalSteps = stepCount - initialStepCount

      val now = System.currentTimeMillis()
      val cadence: Double

      if (cadenceWindowReset) {
        // First step after a stop — start fresh window, use conservative estimate
        cadenceWindowReset = false
        lastStepCount = stepCount
        lastStepTime = now
        cadence = 60.0
      } else {
        val deltaSteps = stepCount - lastStepCount
        val deltaSeconds = (now - lastStepTime) / 1000.0
        cadence = if (deltaSeconds > 0) (deltaSteps / deltaSeconds) * 60.0 else 0.0
        lastStepCount = stepCount
        lastStepTime = now
      }

      // Reset cadence timeout on every real step event
      cadenceHandler.removeCallbacks(cadenceTimeoutRunnable)
      sendEvent("onStepUpdate", mapOf(
        "steps" to totalSteps,
        "cadence" to cadence
      ))

      // Schedule cadence=0 if no new steps arrive within timeout
      cadenceHandler.postDelayed(cadenceTimeoutRunnable, cadenceTimeoutMs)
    }

    override fun onAccuracyChanged(sensor: Sensor?, accuracy: Int) {}
  }

  // MARK: - Module Definition
  override fun definition() = ModuleDefinition {
    Name("RunningKit")

    Events("onLocationUpdate", "onStepUpdate", "onSessionStateChange")

    OnCreate {
      val context = appContext.reactContext ?: return@OnCreate
      fusedLocationClient = LocationServices.getFusedLocationProviderClient(context)
      sensorManager = context.getSystemService(Context.SENSOR_SERVICE) as SensorManager
      stepSensor = sensorManager.getDefaultSensor(Sensor.TYPE_STEP_COUNTER)
    }

    // MARK: - Permissions
    AsyncFunction("requestPermissions") { promise: Promise ->
      val context = appContext.reactContext ?: return@AsyncFunction
      val location = if (
        context.checkSelfPermission(android.Manifest.permission.ACCESS_FINE_LOCATION) ==
        android.content.pm.PackageManager.PERMISSION_GRANTED
      ) "granted" else "denied"

      val motion = if (
        context.checkSelfPermission(android.Manifest.permission.ACTIVITY_RECOGNITION) ==
        android.content.pm.PackageManager.PERMISSION_GRANTED
      ) "granted" else "denied"

      promise.resolve(mapOf("location" to location, "motion" to motion))
    }

    // MARK: - Workout Control
    AsyncFunction("startWorkout") { type: String, promise: Promise ->
      if (sessionState != "idle" && sessionState != "stopped") {
        promise.reject("ERR_WORKOUT_ACTIVE", "A workout is already in progress", null)
        return@AsyncFunction
      }

      totalDistance = 0.0
      lastLocation = null
      totalSteps = 0
      initialStepCount = -1
      lastStepCount = 0
      lastStepTime = 0
      cadenceWindowReset = true
      workoutStartTime = System.currentTimeMillis()
      sessionState = "active"

      startLocationUpdates()
      startStepCounting()

      sendEvent("onSessionStateChange", mapOf("state" to "active"))
      promise.resolve(null)
    }

    Function("pauseWorkout") {
      if (sessionState != "active") return@Function
      sessionState = "paused"
      fusedLocationClient.removeLocationUpdates(locationCallback)
      sensorManager.unregisterListener(stepListener)
      cadenceHandler.removeCallbacks(cadenceTimeoutRunnable)
      sendEvent("onSessionStateChange", mapOf("state" to "paused"))
    }

    Function("resumeWorkout") {
      if (sessionState != "paused" && sessionState != "auto-paused") return@Function
      val wasManuallPaused = sessionState == "paused"
      sessionState = "active"

      startLocationUpdates()

      // Step sensor was stopped during manual pause but kept running during auto-pause
      if (wasManuallPaused) startStepCounting()

      sendEvent("onSessionStateChange", mapOf("state" to "active"))
    }

    // Auto-pause: stops GPS but keeps step sensor running for auto-resume detection
    Function("autoPauseWorkout") {
      if (sessionState != "active") return@Function
      sessionState = "auto-paused"
      fusedLocationClient.removeLocationUpdates(locationCallback)
      // Step sensor intentionally kept running to detect movement for auto-resume
      sendEvent("onSessionStateChange", mapOf("state" to "auto-paused"))
    }

    AsyncFunction("stopWorkout") { promise: Promise ->
      if (sessionState != "active" && sessionState != "paused" && sessionState != "auto-paused") {
        promise.reject("ERR_NO_WORKOUT", "No active workout to stop", null)
        return@AsyncFunction
      }

      fusedLocationClient.removeLocationUpdates(locationCallback)
      sensorManager.unregisterListener(stepListener)
      cadenceHandler.removeCallbacks(cadenceTimeoutRunnable)
      sessionState = "stopped"

      val duration = (System.currentTimeMillis() - workoutStartTime) / 1000.0
      val avgSpeed = if (duration > 0) totalDistance / duration else 0.0
      val calories = totalSteps * 0.04

      sendEvent("onSessionStateChange", mapOf("state" to "stopped"))
      promise.resolve(mapOf(
        "duration" to duration,
        "distance" to totalDistance,
        "steps" to totalSteps,
        "avgSpeed" to avgSpeed,
        "calories" to calories
      ))
    }
  }

  // MARK: - Helpers

  private fun startLocationUpdates() {
    val request = LocationRequest.Builder(Priority.PRIORITY_HIGH_ACCURACY, 1000)
      .setMinUpdateDistanceMeters(0f)
      .build()
    fusedLocationClient.requestLocationUpdates(request, locationCallback, Looper.getMainLooper())
  }

  private fun startStepCounting() {
    stepSensor?.let {
      sensorManager.registerListener(stepListener, it, SensorManager.SENSOR_DELAY_NORMAL)
    }
  }
}
